constraint-solver/src/rule_based_optimizer/environment.rs

```diff
@@ -132,6 +132,41 @@ impl<T: FieldElement> Environment<T> {
         self.single_occurrence_variables.iter()
     }
 
+    /// Split Expr into expr1 + expr_rest, i.e., expr = expr1 + expr_rest
```

Comment:
Also this could be more generic:
Just split off the first summand - something like
```
let head = expr.into_summands().next()?;
let tail = expr - head;
```
But maybe we could even implement this as a function of GroupedExpression for performance reasons.